Framework/Geometry/src/Instrument/Goniometer.cpp

```diff
@@ -192,20 +192,39 @@ void Goniometer::setRotationAngle(size_t axisnumber, double value) {
  * Q Sample
  * @param position :: Q Sample position in reciprocal space
  * @param wavelength :: wavelength
+ * @param flip_x :: option if the q_lab x value should be negative, hence the
+ * detector of the other side of the beam
+ * @param inner :: whether the goniometer is the most inner (phi) or most outer
+ * (omega)
  */
 void Goniometer::calcFromQSampleAndWavelength(
-    const Mantid::Kernel::V3D &position, double wavelength) {
+    const Mantid::Kernel::V3D &position, double wavelength, bool flip_x,
+    bool inner) {
   V3D Q(position);
   if (Kernel::ConfigService::Instance().getString("Q.convention") ==
       "Crystallography")
     Q *= -1;
+
+  Matrix<double> starting_goniometer = getR();
+
+  if (!inner)
+    Q = starting_goniometer * Q;
+
   double wv = 2.0 * M_PI / wavelength;
   double norm_q2 = Q.norm2();
   double theta = acos(1 - norm_q2 / (2 * wv * wv)); // [0, pi]
-  double phi = asin(-Q[1] / wv * sin(theta));       // [-pi/2, pi/2]
+  double phi = asin(-Q[1] / (wv * sin(theta)));     // [-pi/2, pi/2]
```

Comment:
Yikes!

Reply:
Yep, a small bug I found :laughing: This meant it worked fine for anything at ~90 degrees which was fine for HB2C but definitely not for HB3A.